asynch/pool.py

```diff
@@ -141,39 +141,78 @@ async def _create_connection(self) -> None:
+        await conn.connect()
         self._free_connections.append(conn)
 
-    async def _acquire_connection(self) -> Connection:
+    def _pop_connection(self) -> Connection:
         if not self._free_connections:
             raise AsynchPoolError(f"no free connection in {self}")
 
-        conn = self._free_connections.popleft()
+        return self._free_connections.popleft()
+
+    async def _get_fresh_connection(self) -> Optional[Connection]:
+        while self._free_connections:
+            conn = self._pop_connection()
+            try:
+                await conn._refresh()
+                return conn
+            except ConnectionError as e:
+                msg = f"the {conn} is invalidated: {e}"
+                logger.warning(msg)
+        return None
+
+    async def _acquire_connection(self) -> Connection:
+        if conn := await self._get_fresh_connection():
+            self._acquired_connections.append(conn)
+            return conn
+
+        avail = self.maxsize - self.connections
+        if avail < 0:
+            msg = f"no fresh connection to acquire from {self}"
+            raise AsynchPoolError(msg)
+
+        await self._create_connection()
+        conn = self._pop_connection()
         self._acquired_connections.append(conn)
         return conn
 
     async def _release_connection(self, conn: Connection) -> None:
         if conn not in self._acquired_connections:
             raise AsynchPoolError(f"the connection {conn} does not belong to {self}")
 
+        try:
+            await conn._refresh()
+            self._free_connections.append(conn)
+        except ConnectionError:
+            pass  # the invalidated connection is lost
+
         self._acquired_connections.remove(conn)
-        self._free_connections.append(conn)
 
     async def _init_connections(self, n: Optional[int] = None) -> None:
         to_create = n if n is not None else self.minsize
+
         if to_create < 0:
-            msg = f"cannot create ({to_create}) negative connections for {self}"
+            msg = f"cannot create negative number ({to_create}) of connections for {self}"
             raise ValueError(msg)
-        if to_create == 0:
-            return
         if (self.connections + to_create) > self.maxsize:
-            msg = f"cannot create {to_create} connections to exceed the size of {self}"
+            msg = f"cannot create {to_create} connections that will exceed the size of {self}"
             raise AsynchPoolError(msg)
+
+        if not to_create:
+            return
         tasks: list[asyncio.Task] = [
             asyncio.create_task(self._create_connection()) for _ in range(to_create)
         ]
         await asyncio.wait(fs=tasks)
 
+    async def _ensure_minsize_connections(self) -> None:
+        connections = self.connections
+        minsize = self.minsize
+        if connections < minsize:
+            gap = minsize - connections
+            await self._init_connections(min(minsize, gap))
```

Comment:
In which case `gap` can be greater than `minsize`?

Reply:
It is more about that the `_acquire_connection` and `_release_connection` functions may exhaust free and acquired connection deques. So whenever the situation with the (total) number of connections (which is the sum of free and acquired ones)  is less than `pool.minsize`, the gap between the actual number of connections (which is by definition a non-negative integer) and the minimum required by `minsize` should be eliminated.

Reply:
I will simplify the case. You have «a» and «b»:
```py
if a < b:
    c = b - a
    d = min(b, c)
```
What is the point of this? You know exactly that «d» is lower than «b»

Reply:
Nevermind, got it

Reply:
"c" may be equal to "b" and your are right, this `min` may be redundant after all.

Reply:
Yeah, it works without this `min`, thank you, I appreciate.